actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -257,8 +261,19 @@ def non_xhr_javascript_response?
       # * Does the X-CSRF-Token header match the form_authenticity_token
       def verified_request?
         !protect_against_forgery? || request.get? || request.head? ||
-          valid_authenticity_token?(session, form_authenticity_param) ||
-          valid_authenticity_token?(session, request.headers['X-CSRF-Token'])
+          (valid_request_origin? && any_authenticity_token_valid?)
+      end
+
+      # Are any of the authenticity tokens from the request valid?
+      def any_authenticity_token_valid?
+        request_authenticity_tokens.any? do |token|
+          valid_authenticity_token?(session, token)
+        end
+      end
+
+      # Possible authenticity tokens sent in the request.
+      def request_authenticity_tokens
+        [form_authenticity_param, request.headers['X-CSRF-Token']]
```

Comment:
These extra methods make things a bit cleaner and make it easier to override the default Rails behavior. Some apps are trying to move away from using the `X-CSRF-Token` header, and it currently requires quite a bit of hacking to remove.